### DISCUSSION TITLE
Remote backup Solution: Kopia + Tailscale

### DISCUSSION THREAD

**Discussion by radh21301** [immich-app/immich#9868] [Community Guides]
# Remote Backup: Tailscale + Kopia

This is the simplest approach I could find.

To make a remote backup, you need two components:
1. **A backup tool** – in this case, Kopia. 
   - There are other tools like Restic, Borg, etc. However, they don't have a native GUI. I am sure there are other options as well. 
3. **A way to connect to a remote computer** – in this case, Tailscale.
   - I chose Tailscale because it has a feature called Taildrive, which exposes any folder on your computer via WebDAV. Any computers in your Tailscale network can access these folders, even your iPhone! The process to do this is straightforward.
   - The beautiful thing is that Kopia can connect to the WebDAV server directly from the GUI. You don’t have to mount the WebDAV server on your computer first and then direct Kopia to use that mount. I felt this was a very convenient solution for all the noobs like me.

## Setting up Tailscale

- Install Tailscale based on the instructions on the [Tailscale website](https://tailscale.com).
- Enable Taildrive based on the instructions [here](https://tailscale.com/kb/1369/taildrive?tab=linux#enable-taildrive-in-the-policy-file).
  - You can do advanced things like define sharing permissions with other users, etc. But I am the only person using Tailscale, and I don’t need to mess with such things. But there are options if you need them. Mine basically looks like this:
<img width="271" alt="image" src="https://github.com/immich-app/immich/assets/45341381/2c199425-00f1-4d61-9d0b-1d45fe7e6d86">

- Create a folder on the remote computer where you want to store the ‘kopia repository.’
  - Say: `/mnt/user/backups/kopia_repo`

- Share the folder with Taildrive: [Instructions here](https://tailscale.com/kb/1369/taildrive?tab=linux#sharing-and-accessing-folders-with-taildrive).
  - In my case (Linux user):
    ```bash
    # This is the generic command
    tailscale drive share <share-name> <path>
    # This is the command with the details filled in
    tailscale drive share kopia_repo /mnt/user/backups/kopia_repo
    ```
  - That’s it.

## Setting up Kopia

1. Shut down immich before you do the backup to avoid any conflicts.
2. As soon as you launch Kopia, you will see this screen. Click on the WebDAV server.
<img width="468" alt="image" src="https://github.com/immich-app/immich/assets/45341381/fcfe80e2-cd12-451b-8bc5-80d420989064">

4. In the WebDAV server URL enter:
   - `http://100.100.100.100:8080/<user@email.com>/<machine-name>/<share-name>`
   - **Email**: The one you used to sign in to Tailscale.
   - **Machine-name**: The name you gave to that computer in Tailscale.
   - **Share-name**: The name of the share you gave earlier.
   - **Username and password** for the remote computer.

   So, for this example: `http://100.100.100.100:8080/user@email.com/debian-remote/kopia_repo`
<img width="468" alt="image" src="https://github.com/immich-app/immich/assets/45341381/7d79a74f-45cf-44fa-aa04-01273b7fd9e6">

5. The next page asks for a repository password: Enter a strong password.

<img width="468" alt="image" src="https://github.com/immich-app/immich/assets/45341381/69cadc78-f30e-46a0-b786-c38d884ceef0">

6. Then hit "New snapshot."

<img width="468" alt="image" src="https://github.com/immich-app/immich/assets/45341381/a37ce300-4e22-4271-bdbe-6189fa8651d4">

7. Enter the folder where you have all photos/videos from Immich. And under Files (Ignore files) – add the path to the folder where the encoded videos and thumbnails are stored. We don’t really need to back these up. This path is relative to the original path (`/mnt/user/immich-assets`).
<img width="468" alt="image" src="https://github.com/immich-app/immich/assets/45341381/d82e93d7-3856-48a8-ac2e-934b23e68f42">

8. You can add compression if you want.
<img width="468" alt="image" src="https://github.com/immich-app/immich/assets/45341381/76937a3f-5b22-42f8-8a90-c926bcae622c">

9. Once these are done, hit ‘Snapshot now’. You are all done. This might take a while.
10. You can add external_libraries in the same fashion. Start from 'New Snapshot' -Step 5
11. It is best to back up the database folder while you back up the photos and videos (basically add another snapshot for the folder where the Postgres database is located)- Ensure Immich is shut down before you do this.

----------------------
**An automated solution to trigger Kopia backups using cron: Instead of using the schedules from Kopia**
*Advantage: Enables you to stop immich containers before triggering a backup. This is a solid way to do that backup*
```bash
#!/bin/bash
# Stop other immich containers
docker stop immich_machine_learning
docker stop immich_server
docker stop immich_redis

echo "Creating a PostgreSQL DB dump"
FileName=$(date +"%Y-%m-%d-T-%H-%M-%S")
DumpDir="/mnt/user/immich-assets/immich_database_dump"

# Ensure the dump directory exists
mkdir -p "$DumpDir"

# Run the pg_dumpall command without compression (why? refer to this https://immich.app/docs/guides/template-backup-script ) inside the Docker container and compress the output
if docker exec -t immich_postgres pg_dumpall --clean --if-exists --username=postgres > "$DumpDir/${FileName}_dump.sql"; then
  echo "Database dump created successfully: $DumpDir/${FileName}_dump.sql"
else
  echo "Failed to create database dump" >&2
  exit 1
fi

# Run kopia backup: 'Note: my kopia docker container has the mount volume -> /mnt/user/immich-assets:/mnt/user/immich-assets'
docker exec Kopia-remote kopia snapshot create "/mnt/user/immich-assets" # this directory has to be relative to kopia container. It has all images, videos, as well as the immich database dump

# Start the containers back up
docker start immich_redis
docker start immich_server
docker start immich_machine_learning
```

**Comment by StackShard** [comment]
Awesome work, thanks! Now to suss out exactly how to accomplish this on my build... along with some automation to start/stop Immich services as it happens.

**Reply by radh21301** [reply]
I know it has been a while. But I added the script I use with cron to automate the database dump and triggering the backup with kopia.

**Reply by leleogere** [reply]
Note that kopia has a feature to run actions before/after taking a snapshot. I haven't used them yet, but this might be a solution if you want to rely only on kopia's built-in scheduler.

https://kopia.io/docs/advanced/actions

**Reply by radh21301** [reply]
Yeap. I did see that. But sometimes it feels easier to have a bash script that triggers the kopia snapshot. 
Definitely the same result.

**Comment by JoHo94** [comment]
Thanks for the inspiration!

One quote from the official Immich BorgBackup Skript.
`# For deduplicating backup programs such as Borg or Restic, compressing the content can increase backup size by making it harder to deduplicate. If you are using a different program or still prefer to compress, you can use the following command instead`

I think this would be also beneficial here. Kopia can than compress it later :)

https://immich.app/docs/guides/template-backup-script/

**Reply by radh21301** [reply]
Great point. Updated the post.
